src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+
+        public DiagnosticsServerRouter(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public ILogger Logger
+        {
+            get { return _logger; }
+        }
+
+        public virtual void Start()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task Stop()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void Reset()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task<ConnectedRouter> ConnectRouterAsync(CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    /// <summary>
+    /// This class represent a TCP/IP server endpoint used when building up router instances.
+    /// </summary>
+    internal class TcpServerRouter : DiagnosticsServerRouter
+    {
+        protected string _tcpServerAddress;
+
+        protected ReversedDiagnosticsServer _tcpServer;
+        protected IpcEndpointInfo _tcpServerEndpointInfo;
+
+        public int RuntimeInstanceConnectTimeoutMs { get; set; } = 60000;
+        public int TcpServerConnectTimeoutMs { get; set; } = 5000;
+
+        public Guid RuntimeInstanceId
+        {
+            get { return _tcpServerEndpointInfo.RuntimeInstanceCookie; }
+        }
+
+        public int RuntimeProcessId
+        {
+            get { return _tcpServerEndpointInfo.ProcessId; }
+        }
+
+        public string TcpServerAddress
+        {
+            get { return _tcpServerAddress; }
+        }
+
+        protected TcpServerRouter(string tcpServer, int runtimeTimeoutMs, ILogger logger)
+            : base(logger)
+        {
+            _tcpServerAddress = tcpServer;
+            if (string.IsNullOrEmpty(_tcpServerAddress))
+            {
+                _tcpServerAddress = "127.0.0.1:0";
+            }
+            else
+            {
+                _tcpServerAddress = _tcpServerAddress.Replace("tcp://", "");
+                _tcpServerAddress = _tcpServerAddress.Replace(Uri.UriSchemeNetTcp, "");
+            }
+
+            RuntimeInstanceConnectTimeoutMs = runtimeTimeoutMs;
+
+            _tcpServer = new ReversedDiagnosticsServer(_tcpServerAddress, true);
+            _tcpServerEndpointInfo = new IpcEndpointInfo();
+        }
+
+        public override void Start()
+        {
+            _tcpServer.Start();
+        }
+
+        public override async Task Stop()
+        {
+            await _tcpServer.DisposeAsync().ConfigureAwait(false);
+        }
+
+        public override void Reset()
+        {
+            if (_tcpServerEndpointInfo.Endpoint != null)
+            {
+                _tcpServer.RemoveConnection(_tcpServerEndpointInfo.RuntimeInstanceCookie);
+                _tcpServerEndpointInfo = new IpcEndpointInfo();
+            }
+        }
+
+        protected async Task<Stream> ConnectTcpStreamAsync(CancellationToken token)
```

Comment:
I'm a little turned around with this naming scheme. Typically when I see `connect` I assume we're talking about client->server from the client's perspective, but in this case, I think the code is being called from the server's perspective. That makes me think this should be called `AcceptTcpStreamAsync`. Am I misunderstanding the intent here?

Reply:
I will try to clarify this.